common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitSmartcardSession.java

```diff
@@ -0,0 +1,174 @@
+     * @param piv A PivSession created from a SmartCardConnection that can interact with certificates located in the PIV slots on the YubiKey.
+     * @param certList A List collecting the YubiKitCertDetails of the certificates found on the YubiKey.
+     * @throws IOException          in case of connection error
+     * @throws ApduException        in case of an error response from the YubiKey
+     * @throws BadResponseException in case of incorrect YubiKey response
+     */
+    private void getAndPutCertDetailsInList(@NonNull final Slot slot,
+                                            @NonNull final PivSession piv,
+                                            @NonNull final List<ICertDetails> certList)
+            throws IOException, ApduException, BadResponseException {
+        final String methodTag = TAG + ":getAndPutCertDetailsInList";
+        try {
+            final X509Certificate cert =  piv.getCertificate(slot);
+            //If there are no exceptions, add this cert to our certList.
+            certList.add(new YubiKitCertDetails(cert, slot));
+        } catch (final ApduException e) {
+            //If sw is 0x6a82 (27266), This is a FILE_NOT_FOUND error, which we should ignore since this means the slot is merely empty.
+            if (e.getSw() == 0x6a82) {
+                Logger.verbose(methodTag, slot + " slot is empty.");
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * Verifies PIN attempt provided by user.
+     * @param pin Char array with pin.
+     * @return true if PIN is verified; false otherwise.
+     * @throws ApduException in case of an error response from the YubiKey
+     * @throws IOException in case of connection error
+     */
+    @Override
+    public boolean verifyPin(char[] pin) throws ApduException, IOException {
+        try {
+            piv.verifyPin(pin);
+            //If no InvalidPinException is thrown, PIN is validated.
+            return true;
+        } catch (InvalidPinException e) {
+            return false;
+        }
+    }
+
+    /**
+     * Gets number of PIN attempts remaining for YubiKey.
+     * @return number of PIN attempts remaining.
+     * @throws ApduException in case of an error response from the YubiKey
+     * @throws IOException in case of connection error
+     */
+    @Override
+    public int getPinAttemptsRemaining() throws ApduException, IOException {
+        return piv.getPinAttempts();
+    }
+
+    /**
+     * Instantiates a PivPrivateKey to pass to a ClientCertRequest to proceed with certificate based authentication.
+     * @param certDetails YubiKeyCertDetails of chosen certificate.
+     * @param pin Char array containing verified PIN.
+     * @return a PivPrivateKey.
+     * @throws Exception for KeyStore, and if certDetails or PrivateKey are not of correct child type.
+     */
+    @NonNull
+    @Override
+    public PrivateKey getKeyForAuth(ICertDetails certDetails, char[] pin) throws Exception {
```

Comment:
Final? Null check on input?

Reply:
Since `ArrayAdapter.getItem()` could return null, I added a null check after that method is called within `SmartcardCertPickerDialog`. Made parameters in `getKeyForAuth` @Nonnull.